Add ManagedRadioUnitRegistry

matthewkwsiu opened the pull request:
Add abstract class for ManagedRadioUnitRegistry
Add getRadioUnit methods to allow for getting radios based on particular filters
Add addRadioUnit functionality 

TODO: 
- Need to do some testing

Commit:
small typo fix

Commit:
Implementing ConcreteRadioUnit

dLeung09 committed:
Logic for Ericsson-LTE radio and factories added

* Implemented the concrete Ericsson LTE radio unit

* Implemented the factory responsible for creating the Ericsson LTE
radio unit

* Implemented the command executor that links an AbstractRadioUnit to
the concrete Ericsson LTE radio unit

* Implemented the factory responsible for creating the Ericsson LTE
command executor

* Added a simple client for testing behavior added in this commit

Commit:
add some javadocs and reformat order of methods

Commit:
add print statements, modify constructor, edit demo class

Commit:
add abstract class and boilerplate for ManagedRadioUnitRegistry

dLeung09 commented on src/main/java/radiounit/AbstractManagedRadioUnitRegistry.java:
```diff
@@ -0,0 +1,34 @@
+
+	List<ManagedRadioUnit> radioUnits;
+
+	public abstract void addRadioUnit(String ipAddress, String name, Vendor vendor);
```
May want to add the RAT type here

matthewkwsiu replied:
Good point

dLeung09 commented on src/main/java/radiounit/AbstractManagedRadioUnitRegistry.java:
```diff
@@ -0,0 +1,34 @@
+	List<ManagedRadioUnit> radioUnits;
+
+	public abstract void addRadioUnit(String ipAddress, String name, Vendor vendor);
+
```
I like all these methods below! We could look into providing a more sophisticated method, where you pass some combination of vendor, RAT type, alarm status and state, and get back a list of radios that meet all the criteria.

Maybe for now, we could add a method with the following signature that serves that purpose, and potentially revise after:

`public abstract List<ManagedRadioUnit> getRadio(Venor v, RatTypee r, AlarmStatusLevel a, RadioUnitState s);`

matthewkwsiu replied:
Hmm, okay. I think that makes sense. Would the use case be for when you don't recall what the name is?

matthewkwsiu replied:
I clarified this with David

dLeung09 commented on src/main/java/radiounit/ManagedRadioUnitRegistry.java:
```diff
@@ -0,0 +1,64 @@
+ * @author esiumat
+ *
+ */
+public class ManagedRadioUnitRegistry extends AbstractManagedRadioUnitRegistry {
```
We'll need to add a list (or map?) or radios here

matthewkwsiu replied:
Pardon, I'm not super clear on why. Wouldn't this have the list of radios from the abstract class it's extending?

dLeung09 commented on src/main/java/radiounit/ManagedRadioUnitRegistry.java:
```diff
@@ -0,0 +1,64 @@
+ */
+public class ManagedRadioUnitRegistry extends AbstractManagedRadioUnitRegistry {
+
+	public ManagedRadioUnitRegistry() {
```
This should probably be a singleton eventually.

matthewkwsiu replied:
Possibly. Do you think there's a case where we would have multiple clients with multiple registries?

dLeung09 replied:
I don't believe so - the purpose of the singleton is to ensure that there is only one list of radios.

dLeung09 reviewed (commented):
Looks good, not sure if this is the final form for these classes, but added some thoughts regardless

cheeryr commented on src/main/java/radiounit/ManagedRadioUnitRegistry.java:
```diff
@@ -0,0 +1,64 @@
+
+	@Override
+	public List<ManagedRadioUnit> getByState(RadioUnitState state) {
+		// TODO Auto-generated method stub
```
This comment is not relate to registry, but I wander if you can create a RadioUnitState enum? It's more suitable to be passed around as parameter.

matthewkwsiu replied:
Hi, thanks for the comment, and apologies for taking a while to respond. 

Hmm, that's possible. Can you explain your logic a little bit more? I think we had been using the states since we were using the state pattern, and since we already have those state classes, we can reuse them here. 

Are you saying that we could pass Enums as parameters since they'd be easier in terms of memory usage?

cheeryr replied:
It's just more easy for UI(or NMS and mediator) to use it. Because the current getByState method takes a RadioUnitState object as paramenter. To create a RadioUnitState object, we need to pass in a RU object( the constructor of the concrete RadioUnitState class takes a RU as parameter.

It is unrealistic for UI to hold a RU object and then pass in to RadioUnitState to instantiate a RadioUnitState object. A RadioUnitState enum will be more suitable for UI to use as parameter.

You can still keep the state classes but change its name property to state enum.

matthewkwsiu replied:
Hey Yangrui, sorry for not responding for a little while again. 

I think what you said seems to make sense. I don't quite understand the situation fully, but it seems like there is an issue with our current design that cause poor coupling (based on what you said about the UI holding an RU object). 

Perhaps we could call about this sometime next week?

dLeung09 replied:
I don't think this will be that difficult to address -- we can just add an enum that will identify a RadioUnitState class.

The RadioUnitState class can hold a constant final member with this enum (set by the constructor), and provide a getter for it

dLeung09 replied:
Let's merge this one, and address the comment in a future submission

Commit:
Add RatType parameter in addRadioUnit method and add getSpecificRadio

Commit:
add filtered lists of radioUnits

Commit:
Add addRadioUnit functionality

dLeung09 commented on src/main/java/radiounit/ManagedRadioUnitRegistry.java:
```diff
@@ -0,0 +1,86 @@
+	@Override
+	public void addRadioUnit(String ipAddress, String name, Vendor vendor, RatType ratType) {
+		// check that radio hasn't already been added
+		if (getByIpAddress(ipAddress).isEmpty()) {
```
Should this not be `if (!getByIpAddress(ipAddress).isEmpty())`?

dLeung09 commented on src/main/java/radiounit/ManagedRadioUnitRegistry.java:
```diff
@@ -0,0 +1,86 @@
+	}
+
+	@Override
+	public List<ManagedRadioUnit> getSpecificRadio(String ipAddress, String name, Vendor vendor, RatType ratType) {
```
Minor: IP Address should be unique enough that the other parameters don't matter.

Perhaps a more useful method would be one that contains all parameters EXCEPT the IP address...

dLeung09 reviewed (commented):
Looks good - just the one comment where we're checking if the radio already exists - I think the condition needs to be inverted

john-breton committed:
Add required missing functionality

- We need a way to remove ManagedRadioUnits from the registry per the requirements for Demo 2.
- We also need a way to get a ManagedRadioUnit by FrequencyBand per the requirements for Demo 2.
- I also changed the return type of getByIpAddress to be a single ManagedRadioUnit since IPs are unique. There should only ever be one RadioUnit with a specific IP, so there is no need to return it in a List.

john-breton committed:
Add getByBand functionaility

This could probably be a stream but oh well

john-breton committed:
Add getAllRadios() method

john-breton commented:
I've adapted the NetworkManagementSystem/Mediator to be ready to use this class once the code is merged. I'm not sure if there is anything else in this PR that needs to be completed, but I'll wait for it to merge before merging the NMS code.

john-breton committed:
Invert logic for addRadioUnit

matthewkwsiu commented:
Looked over changes with David. Will be addressing unresolved comments in other PRs. Thanks for the help adding the missing functionality John, really appreciate it.
